### PR TITLE
Introduce tests sharding

### PR DESCRIPTION
**Pull request by wzieba**
### Description

This PR implements instrumentation tests sharding using Fladle Gradle Plugin. Comparing to the WooCommerce Android configuration (https://github.com/woocommerce/woocommerce-android/pull/12029) it has some twists:
- support for two different flavors: `wordpress` and `jetpack`
- working Buildkite annotations (see: https://buildkite.com/automattic/wordpress-android/builds/19466#01910d9b-4833-4b77-9a90-92315ee8da3e )

### Impact

Duration of execution of instrumented tests

| | Jetpack | WordPress |
| --- | --- | --- |
| Before (b5a72e383f77d7bdf1cd823166e64793f6514503) | 8m 39s | 6m 17s |
| After | 2m 2s | 2m 9s |
| Diff | -76.5% | -66% |

### Testing

I've rerun the whole job several times to make sure that tests aren't flaky. We shouldn't expect any difference in tests execution though, as we only distribute tests across few devices, instead of running all of them on a single device.

### Demo

image

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: fastlane/lanes/test.rb
Comment (dangermattic) [inline fastlane/lanes/test.rb:22]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/d00c6b66954fd2f8bd460a5df40c7961e7d433e2/fastlane/lanes/test.rb#L22"></span>Layout/IndentationWidth: Use 2 (not 3) spaces for indentation.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/test.rb
Comment (dangermattic) [inline fastlane/lanes/test.rb:24]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/d00c6b66954fd2f8bd460a5df40c7961e7d433e2/fastlane/lanes/test.rb#L24"></span>Layout/RescueEnsureAlignment: <code>rescue</code> at 24, 6 is not aligned with <code>begin</code> at 21, 4.</td>
    </tr>
  </tbody>
</table>

Comment (dangermattic) [inline fastlane/lanes/test.rb:24]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/d00c6b66954fd2f8bd460a5df40c7961e7d433e2/fastlane/lanes/test.rb#L24"></span>Style/RescueStandardError: Avoid rescuing without specifying an error class.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/test.rb
Comment (dangermattic) [inline fastlane/lanes/test.rb:25]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/d00c6b66954fd2f8bd460a5df40c7961e7d433e2/fastlane/lanes/test.rb#L25"></span>Layout/IndentationWidth: Use 2 (not 1) spaces for indentation.</td>
    </tr>
  </tbody>
</table>

File: fastlane/lanes/test.rb
Comment (dangermattic) [inline fastlane/lanes/test.rb]: <table data-meta="generated_by_pr-check">
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td width="100%" data-sticky="false"><span data-href="https://github.com/wordpress-mobile/WordPress-Android/blob/10c9a743b8050a4af2a3beebd928316cb6b25308/fastlane/lanes/test.rb#L24"></span>Style/RescueStandardError: Avoid rescuing without specifying an error class.</td>
    </tr>
  </tbody>
</table>

Comment (wzieba) [reply fastlane/lanes/test.rb]: @iangmaia could you please confirm if it's okay to ignore? I want to rescue from all possible causes of `gradle` or `sh` crash.

Comment (iangmaia) [reply fastlane/lanes/test.rb]: I think those would still be caught if using `rescue StandardError`?

```suggestion
    rescue StandardError
```

By the way, mentioning just in case (not sure if it would make more sense here, as you also have the `gradle` call): for `sh`, you [can also use](https://docs.fastlane.tools/actions/sh/) a `error_callback` parameter if you want to do a custom handling in case the command returns an error.

File: fastlane/lanes/test.rb
Comment (iangmaia) [inline fastlane/lanes/test.rb]: For the `sh` call itself, I'd suggest to break down the command and its arguments (in principle for easier reading, but also for not leaving room for  command injections, issues with spaces in arguments, unwanted interpolation, etc):

```suggestion
      details_url = sh('jq', '.[].webLink', '../build/instrumented-tests/matrix_ids.json', '-r')
```

File: fastlane/lanes/test.rb
Comment (wzieba) [reply fastlane/lanes/test.rb]: True, `rescue StandardError` worked fine, thanks! Addressed in 997e8087676b9c8110fc676501802c1e823edfc1

About `error_callback` - I don't see it being mentioned in `gradle` action documentation, and Gradle part is the most crucial part of this handling so I think we could keep this try/catch approach

File: fastlane/lanes/test.rb
Comment (wzieba) [reply fastlane/lanes/test.rb]: Sure, thanks. Addressed in e7d901b8f318e810a92fcb0c3745600ccf5fe9ec

File: WordPress/build.gradle
Comment (jostnes) [inline WordPress/build.gradle:35]: for test targets unavailable on the variant, do we need to define them here to be excluded? 

asking because when looking at the tests, it looks like there's a check to see if it's jetpack app before continuing with test setup: https://github.com/wordpress-mobile/WordPress-Android/blob/a93b99968d4497ade86305d6e144457dc1170504/WordPress/src/androidTest/java/org/wordpress/android/e2e/StatsGranularTabsTest.kt#L24

Comment (wzieba) [reply WordPress/build.gradle:35]: Unfortunately, they have to. If they're not, `fladle/flank` might create a shard with only tests from `StatsTests` or `StatsGranularTabsTest`. As they're ignored, the execution will be very fast (few seconds) and Firebase Test Lab will mark the test run as failed. It will do this to signal a problem, as too aggressive sharding might bring more unwanted costs. Some more details are available in here the comment description: cc4cc3b . Failed run like this can be found here https://buildkite.com/automattic/wordpress-android/builds/19467#01910e2f-5783-4f74-8897-2725f66ce343

image

File: fastlane/lanes/test.rb
Comment (ParaskP7) [inline fastlane/lanes/test.rb:25]: Praise (❤️): Kudos on making the annotations work, you rock!

PS: Let's do the same for WCAndroid too. 🙏

File: WordPress/build.gradle
Comment (ParaskP7) [inline WordPress/build.gradle]: Minor (🔍): Consider making this a one-liner: `])`

File: WordPress/build.gradle
Comment (ParaskP7) [inline WordPress/build.gradle:45]: Question (❓): You are using `user.home` + `.configure` instead of `project.home` + `.configure-files`. Let's apply this change on WCAndroid too, wdyt? 🤔

File: WordPress/build.gradle
Comment (jostnes) [reply WordPress/build.gradle:35]: I see, thanks for the detailed explanation and linking that commit's comment description! What do you think about adding a comment on the test too? So we would know for a future test that's only available on one variant we should add it to the exclusion list too

File: WordPress/build.gradle
Comment (ParaskP7) [inline WordPress/build.gradle]: Suggestion (💡): I understand that you tried to hit 2 birds with 1 stone, but consider the below alternative instead, IMHO it seems simpler and less error prone:

```groovy
String appPathForVariant(String app) {
    String path = "outputs/apk/${app}Vanilla/debug"
    return new File(layout.buildDirectory.dir(path).get().asFile, '*.apk').absolutePath
}
String testPathForVariant(String app) {
    String path = "outputs/apk/androidTest/${app}Vanilla/debug"
    return new File(layout.buildDirectory.dir(path).get().asFile, '*.apk').absolutePath
}
```

Then, you could just use those as such:

```groovy
        wordpress {
            def app = "wordpress"
            debugApk.set(appPathForVariant(app))
            instrumentationApk.set(testPathForVariant(app))
        }
        jetpack {
            def app = "jetpack"
            debugApk.set(appPathForVariant(app))
            instrumentationApk.set(testPathForVariant(app))
        }
```

The above, to me, it looks a bit more simplified, wdyt? 🤔

**Review by ParaskP7 — APPROVED** [review]
👋 @wzieba !

I have reviewed this PR and everything LGTM, once again, a really awesome job done here, kudos! 🌟 x 🌟 ^ 🌟

1. Thanks for making `fladle` work with a project structure on multiple apps, I am sure it mustn't have been easy! 💯 
2. Thanks for keeping the Firebase annotation! 🥇 

-----

I have left one questions (❓), one suggestions (💡) and one minor (🔍) comment for you to consider. I am going to approve this PR anyway, since none is blocking. I am **NOT** going to merge this PR yet to give you some time to apply any of my suggestions. However, feel free to ignore them and merge the PR yourself.

File: WordPress/build.gradle
Comment (wzieba) [reply WordPress/build.gradle:45]: It's specification of projects, WC and WP/JP has this file's destination configured differently: 
- [WooCommerce](https://github.com/woocommerce/woocommerce-android/blob/trunk/.configure#L43)
- [WordPress/Jetpack](https://github.com/wordpress-mobile/WordPress-Android/blob/trunk/.configure#L48)

File: WordPress/build.gradle
Comment (wzieba) [reply WordPress/build.gradle]: Sure, I'll do this - I wasn't sure what will be the most readable, and was going back and forth with this 😅

File: WordPress/build.gradle
Comment (wzieba) [reply WordPress/build.gradle:35]: Sure thing, added a comment! WDYT? 

9c05ddd98f2b814921613ea56dbf3d57cc60ca2f

File: WordPress/build.gradle
Comment (ParaskP7) [reply WordPress/build.gradle:45]: Ok, cool, so since both those projects have a different `destination`, we should use that and adhere to in on the `fladle` side too, gotcha, I was just trying to make both project consistent, but this seems a bit more work than I expected... 🫣

File: WordPress/build.gradle
Comment (ParaskP7) [reply WordPress/build.gradle]: Yea, I was reading all the inverted ifs/throws and more if logic, and couldn't help myself for trying to figure out if we can make this a bit simpler... 🫣 To me, less ifs, less complexity... 😅

**Pull request merged**
